Add topic availability safety check to controlled shutdown

ambroff opened the pull request:
The controller will not allow a broker to initiate controlled shutdown unless the shutdown is unlikely to lead to a topic-partition ISR shrinking below the min.insync.replicas setting for that topic.

This new functionality is optional and disabled by default. The following new configs have been added:

* controlled.shutdown.safety.check.enable: Enable this safety check for any ControlledShutdownRequest handled by the controller.
* controlled.shutdown.safety.check.redundancy.factor: Additional replicas above min.insync.replicas that are required for any topic-partition to allow shutdown of a broker. Default is 2, so if a topic has min.insync.replicas=2, then the ISR must have 4 replicas to allow shutdown.
* controlled.shutdown.topic.cache.size: In order to implement this safety check the controller needs access to topic configuration. Looking this up requires a read from ZooKeeper, and these configs change infrequently, so caching this will vastly reduce the amount of time required to perform this safety check on large clusters with tens of thousands of topics.
* controlled.shutdown.topic.config.cache.ttl.ms: The number of milliseconds to cache topic configs for this safety check. Default is one minute.

By default a broker attempting shutdown will only send ControlledShutdownRequest to the controller 3 times with a 5 second sleep in between. If it is desired to never skip controlled shutdown, controlled.shutdown.max.retries can be increased to a much larger value than 3. It can be set to 2147483647 (Integer.MAX_VALUE) to effectively retry forever.

A controller integration test has been added which exercises this code with a simulated three node cluster with a topic configured with min.insync.replicas=2, allowing only one broker to shut down at a time.

org.apache.kafka.common.cache.BoundedTimedCache was added to implement a time-based bounded cache for topic configuration in the controller.

In ControllerContext, instead of maintaining a list of broker ids that have begun shutting down called shuttingDownBrokers, a map of brokerId -> brokerEpoch. This will make it easier to keep this map and broker presence notifications (via zookeeper) in sync.

When a ControlledShutdown request is allowed to begin, it is recorded in zookeeper under /brokers/shutdown/, with the value of the broker epoch. On controller initialization this is loaded into controllerContext.shuttingDownBrokers. When a broker has restarted, it will come back with a new broker epoch which will take precedence over the epoch in controllerContext.shuttingDownBrokers.

ambroff committed:
Add topic availability safety check to controlled shutdown

The controller will not allow a broker to initiate controlled shutdown
unless the shutdown is unlikely to lead to a topic-partition ISR
shrinking below the min.insync.replicas setting for that topic.

This new functionality is optional and disabled by default. The
following new configs have been added:

* controlled.shutdown.safety.check.enable: Enable this safety check
  for any ControlledShutdownRequest handled by the controller.
* controlled.shutdown.topic.cache.size: In order to implement this
  safety check the controller needs access to topic
  configuration. Looking this up requires a read from ZooKeeper, and
  these configs change infrequently, so caching this will vastly
  reduce the amount of time required to perform this safety check on
  large clusters with tens of thousands of topics.
* controlled.shutdown.topic.config.cache.ttl.ms: The number of
  milliseconds to cache topic configs for this safety check. Default
  is one minute.

By default a broker attempting shutdown will only send
ControlledShutdownRequest to the controller 3 times with a 5 second
sleep in between. If it is desired to never skip controlled shutdown,
controlled.shutdown.max.retries can be increased to a much larger
value than 3. It can be set to 2147483647 (Integer.MAX_VALUE) to
effectively retry forever.

A controller integration test has been added which exercises this code
with a simulated three node cluster with a topic configured with
min.insync.replicas=2, allowing only one broker to shut down at a
time.

org.apache.kafka.common.cache.BoundedTimedCache was added to implement
a time-based bounded cache for topic configuration in the controller.

In ControllerContext, instead of maintaining a list of broker ids that
have begun shutting down called shuttingDownBrokers, a map of brokerId
-> brokerEpoch. This will make it easier to keep this map and broker
presence notifications (via zookeeper) in sync.

When a ControlledShutdown request is allowed to begin, it is recorded
in zookeeper under /brokers/shutdown/<broker-id>, with the value of
the broker epoch. On controller initialization this is loaded into
controllerContext.shuttingDownBrokers. When a broker has restarted, it
will come back with a new broker epoch which will take precedence over
the epoch in controllerContext.shuttingDownBrokers.

ambroff commented:
This functionality is intended to replace the shutdown safety check we have in our internal Kafka deployment. In that system, we:

1. Acquire a zookeeper lock for the current rack (any other member of that rack will also hold this lock so they can all shut down concurrently)
2. Gather topic metadata to determine partition assignments
3. For each local topic, get topic config to learn min.insync.replicas setting
4. If shutting down will not leave any topic-partitions with ISR smaller than min.insync.replicas, proceed by calling KafkaServer#shutdown

ambroff commented:
I haven't done a lot of testing with this yet. I'm putting it up here as a RFC for now.

ambroff committed:
Add redundancy factor config property

After re-reviewing our internal shutdown check code, I realized that
this code doesn't exactly match. We have an additional redundancy
factor (defaulting to 2), and only allow shutdown if
|isr| >= min.insync.replicas + redundancy factor.

This adds an equivalent redundancy factor config property, and uses
the same condition.

To make the integration test work I changed the topic to be RF=4,
min.insync.replicas=2. The test still shows that only one broker is
allowed to shutdown.

ambroff committed:
Add additional test to ensure controller move is handled

ambroff commented:
One thing I can think of that is missing here is the ability to tell a broker to bypass controlled shutdown at runtime. That is something our internal shutdown check allows and it is required from time to time.

ambroff committed:
Fix build of org.apache.kafka.jmh.server.HighwatermarkCheckpointBench

ambroff committed:
Bugfix. liveBrokerIds is a view on top of liveBrokerIdAndEpochs and shuttingDownBrokerIds that merges on epoch.

ambroff committed:
Delete commented out code I left behind.

gitlw commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -61,7 +61,7 @@ case class ReplicaAssignment(replicas: Seq[Int],
   val stats = new ControllerStats
   var offlinePartitionCount = 0
-  var shuttingDownBrokerIds: mutable.Set[Int] = mutable.Set.empty
+  var shuttingDownBrokerIds: Map[Int, Long] = Map.empty
```
Currently we only maintain the current epoch of each broker in the liveBrokerEpochs.
This PR suggests the epoch in this map may become lagging behind the epochs maintained in the liveBrokerEpochs. Can you please give an example sequence of events that can cause that condition?

ambroff replied:
This map just contains the epoch of each broker when it had last initialized shutdown via the controlled shutdown protocol. So it's OK for it to lag behind the actual epoch of live brokers.

The idea is that ControllerContext#liveBrokerIds will contain the set of broker ids that are considered live. A broker that has been allowed to initiate shutdown is not considered live for the uses of this function.

It is defined as:

```
def liveBrokerIds: Set[Int] = liveBrokerEpochs.filter(b => b._2 > (shuttingDownBrokerIds.getOrElse(b._1, -1): Long)).keySet
```

So take the (brokerId, epoch) for every known live broker, and filter out all brokers which have been marked as shutting down with that brokerEpoch.

So say you start with brokerId 3, epoch 12:

  liveBrokerEpochs contains 3 -> 12
  shuttindDownBrokerIds is empty

If broker 3 is allowed to shut down, then shuttingDownBrokerIds has 3 -> 12 added to it. Now the state is


  liveBrokerEpochs contains 3 -> 12
  shuttingDownBrokerIds contains 3 -> 12

With that state, liveBrokerIds evaluates to the empty set.

When broker 3 comes back up with epoch 13 the state will be 

  liveBrokerEpochs contains 3 -> 13
  shuttingDownBrokerIds contains 3 -> 12

Now since shuttingDownBrokerIds has a the epoch for the previous broker process with id 3, liveBrokerIds will evaluate to the set `{3}`

So it's fine for shuttingDownBrokerIds to have old broker ids. It is just a marker to inform the cluster the epoch at which a given broker had initiated shutdown.

gitlw replied:
Thanks for the example. It's still not clear to me why we need to keep track of the epoch of when a broker initiated the controlled shutdown. What would be missing if we don't?

ambroff replied:
The missing detail is controller movement.

The existing controlled shutdown logic before this patch doesn't keep track of the shutdown epoch because the only reason it cares about the epoch at all is to filter out stale ControlledShutdownRequests that come in for a broker that has already been restarted.

```
    // broker epoch in the request is unknown if the controller hasn't been upgraded to use KIP-380
    // so we will keep the previous behavior and don't reject the request
    if (brokerEpoch != AbstractControlRequest.UNKNOWN_BROKER_EPOCH) {
      val cachedBrokerEpoch = controllerContext.liveBrokerIdAndEpochs(id)
      if (brokerEpoch < cachedBrokerEpoch) {
        val stateBrokerEpochErrorMessage = "Received controlled shutdown request from an old broker epoch " +
          s"$brokerEpoch for broker $id. Current broker epoch is $cachedBrokerEpoch."
        info(stateBrokerEpochErrorMessage)
        throw new StaleBrokerEpochException(stateBrokerEpochErrorMessage)
      }
    }
```

After that block it no longer cares about the epoch and discards that information, only storing the broker id of a shutting down broker into controllerContext.shuttingDownBrokers, which is a Set[Int].

After this patch, the controller is keeping track of which brokers have initiated shutdown so that it doesn't allow multiple brokers to shut down that will have an impact on the availability of a topic-partition they have in common. So it needs a short term history of shutdown requests.

The existing shuttingDownBrokers: Set[Int] could work fine for this, however it doesn't deal with controller movement.

Imagine this hypothetical scenario:
1. You have controller running on broker id 0
1. You have two brokers 2 and 3, and a topic-partition topic-0 with ISR [1, 2].
2. Broker id 2 sends ControlledShutdownRequest to controller 0. Controller does safety check and says 👍 go ahead and shut down.
3. Broker 2 gets ControlledShutdownResponse and begins shutdown
4. Broker 0 crashes and controller moves to broker 1
5. Broker 3 sends ControlledShutdownREquest to controller 1.
6. Controller 1 has no idea that broker 2 is shutting down. Controller does safety check and says 👍  go ahead broker 3, shut down
7. Broker 2 shuts down
8. Broker 3 shuts down, leaving topic-0 offline.

So, that means that in order to handle controller movement, you need the controller to pass state along to the next controller so that it knows which brokers are pending shutdown.

You could do that just by writing the broker id of the brokers pending shutdown to zookeeper, effectively persisting shuttingDownBrokers: Set[Int] to zookeeper.

The problem is, when a new controller boots up after a controller move, then it would only have the broker ids of brokers who had initiated shutdown, without knowing whether that broker had already shut down and started back up again.

Storing brokerId => brokerEpoch in zookeeper under /brokers/shutdown means that when the new controller takes over, it will know right away whether that state passed from the previous controller is stale shutdown information.

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1278,12 +1322,19 @@ class KafkaController(val config: KafkaConfig,
-    controllerContext.shuttingDownBrokerIds.add(id)
+    info(s"Shutting down broker $id")
+
+    zkClient.recordBrokerShutdown(id, brokerEpoch, controllerContext.epochZkVersion)
```
Is there a need to remove the `/brokers/shutdown/<id>` znode when the shutdown is complete?

xiowu0 replied:
I have the same question. 
There are two cases: 
the shutting down broker comes back online. 
the shutting down broker is removed from the cluster. 

should the controller clean the znode in these two cases. 

when the controller detects a broker is offline, can the controller clear the shutting down broker node for it?
When a broker becomes the new controller, it can also clear shutting down broker znodes for nodes that were already offline?

ambroff replied:
I don't think there is any need to remove these znodes, since they only serve as a marker that lets the controller always know the last epoch of a broker when it initiated shutdown. What is the concern with leaving these znodes as they are?

If you think it's important to remove them we can, we'd just have to evaluate when it is safe to do so. I don't think `/brokers/shutdown/<id>` when shutdown is complete is right, but I think you are right xiowu0 that removing when the controller detects that the broker is offline would work. It might be better to remove when the broker comes online with a larger epoch value and liveBrokerEpochs is updated.

ambroff replied:
I think I liked just keeping a map of all shutdown epochs forever because it seemed simplest to get right, but if you feel strongly we should remove them we can figure that out.

ambroff replied:
I pushed a change to remove these znodes in `kafka.controller.KafkaController#onBrokerStartup`

ambroff commented:
Thanks for reviewing guys. I have done some testing locally and on one of our test clusters with a snapshot and this seems to work correctly from my testing.

ambroff commented:
Another important thing to note: This patch depends on interbroker-protocol-version 2.4, since that's the version that added brokerEpoch to ControlledShutdownRequest.

With an earlier IBP version the brokerEpoch is -1 which breaks this code, since the mechanism for determining the set of live brokers (which is the set difference of the set of shutting down brokers) will just always include all live brokers. I guess it could be set to the current known broker epoch of that brokerId instead.

It looks like we don't use 2.4 everywhere in production yet so it is probably best to make that a prerequisite of this patch.

ambroff committed:
Handle case where inter-broker-protocol < 2.4 is used.

ambroff commented:
I updated this to deal with IBP < 2.4 by just using the broker epoch from liveBrokerIdAndEpochs (with a log message).

ambroff committed:
Fix test flakiness.

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1261,6 +1267,44 @@ class KafkaController(val config: KafkaConfig,
+        .toString.toInt
+
+      // See which replicas are known alive and not pending shutdown for this partition
+      val liveReplicas = controllerContext.partitionReplicaAssignment(partition).count( { replicaBrokerId =>
```
An alive broker could still be out of the ISR. I guess we need to check the latest ISR state.
Also since the info in partitionLeadershipInfo could go stale for 10 - 15 seconds, it's safer to query the partitions' ISR state here.

ambroff replied:
Ah you're right this isn't correct. Thanks.

ambroff replied:
I need to focus on OKR planning but I'll fix this later today.

ambroff replied:
I think this is fixed now.

ambroff committed:
Remove /brokers/shutdown/<broker-id> znodes when <broker-id> restarts

ambroff committed:
Feedback from Xiongqi

Broker epoch can change any time the zookeeper session is
reestablished. In order to prevent funny business,
KafkaServer#doControlledShutdown will record its epoch at the time it
sends ControlledShutdownRequest to the controller, and will reject the
response if its epoch has changed by the time it receives the
ControlledShutdownResponse.

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1261,6 +1270,44 @@ class KafkaController(val config: KafkaConfig,
+
+    val atRiskPartitions = controllerContext.partitionsOnBroker(id).filter { partition =>
+      // Look up minISR for this topic, or use the default if not configured.
+      val minISR: Int = fetchTopicConfig(partition.topic())
```
This still fetches each topic's configs sequentially. When the topicConfigCache is cold, and a given broker is likely hosting a partition for every topic, this could possibly take a very long time and thus choke the controller.

ambroff replied:
Yeah I was worried about that. What do you think we should do to fix this?

One option is to scan and watch all topic configuration controller startup. That way at least it doesn't pause while the controller is handling requests.

Another I guess is to have a thread pool for gathering topic configuration to add some parallelism, which would speed this up for the current code where it is done in the ControlledShutdownRequest handler *or* during controller startup.

Another thing is that we don't necessarily have to cache all topic configuration in memory like this. We could just store the min.insync.replicas setting in a map.

xiowu0 replied:
DynamicConfigManager keeps check of topic config changes. could we leverage DynamicConfigManager here so that whenever there is a topic config change,  kafka controller can be updated for the change.

ambroff replied:
Oh that makes this really easy, thanks xiowu0.

ambroff replied:
I updated TopicConfigHandler (owned by DynamicConfigManager) just update a map of topic -> min.insync.replicas in ControllerContext and deleted the cache.

ambroff committed:
Feedback from Lucas

safeToShutdown mistakenly only considered broker liveness, not whether
a broker was in the ISR for a given topic-partition.

This now only considers a replica "live" if it is in the
ISR (according to state the controller has at that moment) and the
replica is "live", meaning that the controller hasn't been notified
that the replicas zookeeper session has expired, and that replica
hasn't begun shutting down.

ambroff committed:
Feedback from Xiongqi and Lucas

Just use DynamicConfigManager for pushing topic config information
into the controller.

* Added a map of topic => min.insync.replicas in ControllerContext
* TopicConfigHandler now calls KafkaController#setMinInSyncReplicas,
  which updates that map

Removed BoundedTimedCache and related configs since they aren't needed
anymore.

ambroff committed:
Remove unecessary whitespace and import changes.

ambroff commented:
The one thing not addressed by this patch is how to skip the shutdown check. In our internal one we can do it via JMX. I'd like to have a better, more secure way to do it here.

I think if we change KafkaServer#doControlledShutdown slightly then we can just do a dynamic broker config instead.

```
diff --git a/core/src/main/scala/kafka/server/KafkaServer.scala b/core/src/main/scala/kafka/server/KafkaServer.scala
index cfcca2d4f271..910a1909a59a 100755
--- a/core/src/main/scala/kafka/server/KafkaServer.scala
+++ b/core/src/main/scala/kafka/server/KafkaServer.scala
@@ -471,7 +471,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
 
     val socketTimeoutMs = config.controllerSocketTimeoutMs
 
-    def doControlledShutdown(retries: Int): Boolean = {
+    def doControlledShutdown(): Boolean = {
       val metadataUpdater = new ManualMetadataUpdater()
       val networkClient = {
         val channelBuilder = ChannelBuilders.clientChannelBuilder(
@@ -514,12 +514,12 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
 
       try {
 
-        var remainingRetries = retries
+        var attempts = 0
         var prevController: Broker = null
         var ioException = false
 
-        while (!shutdownSucceeded && remainingRetries > 0) {
-          remainingRetries = remainingRetries - 1
+        while (!shutdownSucceeded && attempts < config.controlledShutdownMaxRetries) {
+          attempts += 1
 
           // 1. Find the controller and establish a connection to it.
 
@@ -617,7 +617,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
 
       brokerState.newState(PendingControlledShutdown)
 
-      val shutdownSucceeded = doControlledShutdown(config.controlledShutdownMaxRetries.intValue)
+      val shutdownSucceeded = doControlledShutdown()
 
       if (!shutdownSucceeded)
         warn("Proceeding to do an unclean shutdown as all the controlled shutdown attempts failed")
```

Then to skip shutdown on broker 0, we'd just do

```
$ bin/kafka-configs --bootstrap-server ... --entity-type brokers --entity-name 0 --alter --add-config controlled.shutdown.max.retries=0
```

I'll test this after lunch today.

ambroff committed:
No need for this to be a var.

ambroff commented:
> The one thing not addressed by this patch is how to skip the shutdown check. In our internal one we can do it via JMX. I'd like to have a better, more secure way to do it here.
> ...
> I'll test this after lunch today.

Turns out that only works if we convert the config to a dynamic one. I see how it works now. So maybe "controlled.shutdown.enable" could be made into a dynamic broker config.

An alternative would be to do something to get the controller to just skip the safety check. We could make `controlled.shutdown.safety.check.enable` into a dynamic config. But ideally we would make it so that we can tell the cluster to skip the check for an individual broker. So it might be better to inject a command into the controller telling it to skip for a certain (brokerId, epoch).

I think telling the controller to skip shutdown safety check for (brokerId, epoch) is the best thing to do, but it would be pretty lame to introduce a new request type just for this. And it would be even lamer to make twiddling znodes the way for the admin to do it.

gitlw commented:
> > The one thing not addressed by this patch is how to skip the shutdown check. In our internal one we can do it via JMX. I'd like to have a better, more secure way to do it here.
> > ...
> > I'll test this after lunch today.
> 
> Turns out that only works if we convert the config to a dynamic one. I see how it works now. So maybe "controlled.shutdown.enable" could be made into a dynamic broker config.
> 
> An alternative would be to do something to get the controller to just skip the safety check. We could make `controlled.shutdown.safety.check.enable` into a dynamic config. But ideally we would make it so that we can tell the cluster to skip the check for an individual broker. So it might be better to inject a command into the controller telling it to skip for a certain (brokerId, epoch).
> 
> I think telling the controller to skip shutdown safety check for (brokerId, epoch) is the best thing to do, but it would be pretty lame to introduce a new request type just for this. And it would be even lamer to make twiddling znodes the way for the admin to do it.

I like the idea that we add a dynamic broker-level config `controlled.shutdown.safety.check.enable` or `controlled.shutdown.safety.check.disable` to influence the flag on specific brokers:
`./bin/kafka-configs.sh --bootstrap-server localhost:9092 --command-config ssl_configs.properties --alter --entity-type brokers --entity app0.linkedin.com --add-config 'controlled.shutdown.safety.check.disable=true'`

And during shutdown check, each broker encodes its local flag into the ControlleredShutdownRequest before sending it to the controller. What do you think?

ambroff committed:
Add the ability to skip the shutdown safety check

This patch adds LiControlledShutdownSkipSafetyCheckRequest. It takes a
brokerId and brokerEpoch. When sent to the controller, the controller
will record it in a map in memory.

After handling that request, the controller will ignore the results of
the safety check for any request with that brokerId and epoch <= the
epoch provided via LiControlledShutdownSkipSafetyCheckRequest.

This is not persisted to zookeeper since this is an emergency
action. If the controller moves then an admin can just send the
request again.

This request is authorized as a cluster action.

Since LiControlledShutdownSkipSafetyCheck is not submitted to
upstream, it has been added with an ApiKey of 1000. The largest ApiKey
in upstream is 47. This new API is only meant to be used by
AdminClient and is not meant to be an inter-broker request. So this
sets the precedent that if we have an admin-only API we want to add
that isn't upstream yet, we can give it an ApiKey of 1000+. Then if
this internal-only admin API does make it upstream, migrating our
tools to it will be easy since usage should be minimal.

This required fixing a couple of tests that made assumptions about
ApiKeys, namely that the array of api keys now has gaps in it. Now
ApiKeys#hasKeys returns true if the key is within range *and* the
entry in the ID_TO_TYPE array is not null.

ambroff commented:
I just pushed a change that will address the remaining issue, which is to allow admins to skip the safety check.

Lucas and I discussed this a bit and decided that it's best to just add a new RPC request type instead of modifying ControlledShutdownRequest. This is mainly just because it may take some time for upstream to accept a patch like this, and if we have to carry it for a while, merging changes to ControlledShutdownRequest from upstream may be much more difficult in terms of wire format compatibility then it will be to maintain additional request types in our fork.

ambroff committed:
Fix a bad default in TestUtils that I added which broke some tests.

ambroff committed:
Fix KafkaConfigTest. Apparently it fails when you add new config properties.

ambroff committed:
Fix remaining tests

RequestQuotaTests failed after I added
ApiKeys.LI_CONTROLLED_SHUTDOWN_SKIP_SAFETY_CHECK. That's because the
tests need updating in a few places to know how to perform some tests
with that type, but also since I forgot to mark that request type as a
cluster action.

ambroff committed:
Cluster actions should not have throttle time in the response

According to ApiKeysTest anyway.

gitlw commented on clients/src/main/java/org/apache/kafka/clients/admin/SkipShutdownSafetyCheckResult.java:
```diff
@@ -0,0 +1,46 @@
+        this.future = future;
+    }
+
+    public void all() {
```
Maybe expose the future like other *Result classes (e.g. CreateTopicsResult), just in case the caller needs an async API.

ambroff replied:
Sure I just changed that method to return the future.

I'm not sure if all() is the best method name but I'm just following convention.

gitlw commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -287,7 +292,15 @@ public static ApiKeys forId(int id) {
 
     public static boolean hasId(int id) {
-        return id >= MIN_API_KEY && id <= MAX_API_KEY;
+        if (id < MIN_API_KEY) {
```
minor: combine the two if statements with logical or || .

ambroff replied:
I actually feel like it's easier to read as is.

gitlw commented on clients/src/main/java/org/apache/kafka/common/requests/LiControlledShutdownSkipSafetyCheckRequest.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private final LiControlledShutdownSkipSafetyCheckRequestData data;
+    private final short version;
```
It seems there is no need to keep the version var here since the AbstractRequest already has a version field, and can be accessed using the version() method.

ambroff replied:
Sure, fixed.

gitlw commented on clients/src/main/java/org/apache/kafka/common/requests/LiControlledShutdownSkipSafetyCheckRequest.java:
```diff
@@ -0,0 +1,82 @@
+    public static class Builder extends AbstractRequest.Builder<LiControlledShutdownSkipSafetyCheckRequest> {
+        private final LiControlledShutdownSkipSafetyCheckRequestData data;
+
+        public Builder(LiControlledShutdownSkipSafetyCheckRequestData data, short desiredVersion) {
```
minor: rename argument to allowedVersion to align with the parent class API.

ambroff replied:
+1

gitlw commented on clients/src/main/java/org/apache/kafka/common/requests/LiControlledShutdownSkipSafetyCheckResponse.java:
```diff
@@ -0,0 +1,74 @@
+        return data;
+    }
+
+    public static LiControlledShutdownSkipSafetyCheckResponse parse(ByteBuffer buffer, short version) {
```
It seems this method is not being used.

ambroff replied:
I only added it because I was using ControlledShutdownRequest as a template and I wasn't sure whether that was some implicit part of the interface. I'll try removing it.

gitlw commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -61,7 +61,8 @@ case class ReplicaAssignment(replicas: Seq[Int],
-  var shuttingDownBrokerIds: mutable.Set[Int] = mutable.Set.empty
+  var shuttingDownBrokerIds: Map[Int, Long] = Map.empty
+  var skipShutdownSafetyCheck: Map[Int, Long] = Map.empty
   private var liveBrokers: Set[Broker] = Set.empty
```
By reading the logics, I feel the liveBrokerEpochs could be better named as allBrokerEpochs or knownBrokerEpochs, since members in it are not necessarily alive. But that's independent of your change here, and we can perhaps change that in upstream. And the same goes for the liveBrokers field.

gitlw commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -61,7 +61,8 @@ case class ReplicaAssignment(replicas: Seq[Int],
   val stats = new ControllerStats
   var offlinePartitionCount = 0
-  var shuttingDownBrokerIds: mutable.Set[Int] = mutable.Set.empty
+  var shuttingDownBrokerIds: Map[Int, Long] = Map.empty
```
Since this field is being updated, maybe using a mutable.Map could be more efficient?
On the other hand, supposedly the size of this map is small and mostly < 1000, so it's up to you.

ambroff replied:
Yeah you're probably right. I changed it to mutable map just in case.

gitlw commented on core/src/main/scala/kafka/controller/ControllerContext.scala:
```diff
@@ -198,7 +206,7 @@ class ControllerContext {
 
   // getter
-  def liveBrokerIds: Set[Int] = liveBrokerEpochs.keySet -- shuttingDownBrokerIds
+  def liveBrokerIds: Set[Int] = liveBrokerEpochs.filter(b => b._2 > (shuttingDownBrokerIds.getOrElse(b._1, -1): Long)).keySet
```
minor: remove the :Long and use -1L?

ambroff replied:
+1

gitlw commented on core/src/main/scala/kafka/controller/Election.scala:
```diff
@@ -157,7 +157,7 @@ object Election extends Logging {
     val shuttingDownBrokerIds = controllerContext.shuttingDownBrokerIds.toSet
     leaderAndIsrs.map { case (partition, leaderAndIsr) =>
-      leaderForControlledShutdown(partition, leaderAndIsr, shuttingDownBrokerIds, controllerContext)
+      leaderForControlledShutdown(partition, leaderAndIsr, shuttingDownBrokerIds.map(_._1).toSet, controllerContext)
```
move the map(_._1).toSet out of the loop to save some computing.
Also it seems we can use shuttingDownBrokerIds.keySet.toSet

ambroff replied:
Good point.

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1261,6 +1279,51 @@ class KafkaController(val config: KafkaConfig,
+    controllerContext.skipShutdownSafetyCheck += (id -> brokerEpoch)
+  }
+
+  private def safeToShutdown(id: Int, brokerEpoch: Long): Boolean = {
```
I feel we could move the skip check inside this method, but it's up to you.

ambroff replied:
I like having it outside of safeToShutdown since 1.) it makes safeToShutdown a little simpler, and 2.) I want to always run the safety check even if it is skipped so that we have a clear record in the log for cases where it was not safe to shut down but we skipped it anyway.

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1261,6 +1279,51 @@ class KafkaController(val config: KafkaConfig,
+
+      // See which replicas are known alive and not pending shutdown for this partition
+      val liveReplicasInIsr = controllerContext.partitionLeadershipInfo(partition).leaderAndIsr.isr.count({ replicaBrokerId =>
+        controllerContext.liveBrokerIds.contains(replicaBrokerId)
```
save the liveBrokersIds to a val and move it above the line
  controllerContext.partitionsOnBroker(id).filter { partition =>

to save some computing.

ambroff replied:
+1

gitlw commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1261,6 +1279,51 @@ class KafkaController(val config: KafkaConfig,
+      })
+
+      // Consider this topic-partition at-risk if removing one broker will result in the ISR shrinking below minISR
+      debug(s"$partition has min.insync.replicas=$minISR and a redundancy factor of ${config.controlledShutdownSafetyCheckRedundancyFactor}. Broker $id is a replica and the ISR contains $liveReplicasInIsr.")
```
minor: Since the liveRTeplicasInIsr is a count, the log 
    ... contains $liveReplicasInIsr 
isn't clear.

ambroff replied:
Good point, fixed.

gitlw commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -464,6 +464,45 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
+    getChildren(BrokerShutdownNode.path)
+      .map(_.toInt)
+      .map(brokerId => {
+        val getDataRequest = GetDataRequest(BrokerShutdownIdZNode.path(brokerId))
```
can we batch the requests as done in the getAllBrokerAndEpochsInCluster method?

ambroff replied:
Sure that's a good idea.

gitlw commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -464,6 +464,45 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
+    }
+  }
+
+  def removeBrokerShutdown(brokerId: Int, expectedControllerEpochZkVersion: Int): Unit = {
```
also it'd be good if we could batch the remove request to include multiple brokers that need to be removed.

gitlw commented on core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:
```diff
@@ -535,6 +536,103 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
+    assertEquals(0, partitionsRemaining.size)
+    var partitionStateInfo = activeServers.head.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get
+    var leaderAfterShutdown = partitionStateInfo.leader
+    assertEquals(0, leaderAfterShutdown)
```
it seems the leader doesn't have to be broker 0.

gitlw commented on core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:
```diff
@@ -535,6 +536,103 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
+
+    TestUtils.waitUntilTrue(() => notEnoughReplicasDetected, "Fail to detect expected NotEnoughReplicasException")
+
+    val expectedShutdownEntries = Map(2 -> 49)
```
Can you please explain why it'd be epoch 49?

ambroff replied:
That's the epoch every time. I think it's just because this is a fresh zookeeper with a fixed number of clients that are connecting in the same order every time, we end up with the same brokerEpoch (which comes from `kafka.zk.KafkaZkClient#registerBroker`).

If I run this test with logging turned up you can see the brokers register in the same order every time, and that is the czxid from each registration.

```
[2021-01-26 21:37:37,580] INFO Registered broker 3 at path /brokers/ids/3 with addresses: ArrayBuffer(EndPoint(localhost,44075,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 27 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:37,874] INFO Registered broker 2 at path /brokers/ids/2 with addresses: ArrayBuffer(EndPoint(localhost,45451,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 49 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:37,966] INFO Registered broker 1 at path /brokers/ids/1 with addresses: ArrayBuffer(EndPoint(localhost,33383,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 69 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:38,040] INFO Registered broker 0 at path /brokers/ids/0 with addresses: ArrayBuffer(EndPoint(localhost,38473,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 89 (kafka.zk.KafkaZkClient:74)
```

Note that they are registered in reverse order from 3 to 0. The test reverses the server config list before calling createServer on each config.

Beyond that explaining why the epoch is always that is a zookeeper implementation detail.

gitlw replied:
I see and thanks for the explanation. Then can you please put some comment in the code or add a link to this comment above?

ambroff replied:
Done.

gitlw commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -88,6 +88,11 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
       kafkaController.enableTopicUncleanLeaderElection(topic)
     }
+
+    Try(topicConfig.getProperty(KafkaConfig.MinInSyncReplicasProp).toInt) match {
```
It seems if the brokers are already running, and a new topic is created, this code path won't be triggered.
It's only triggered when the config of an existing topic's config is being changed.

Then it'd be a problem for the controller's topicMinIsrConfig field since it won't contain data of the newly created topics. Maybe we'll need to query zk for topic configs upon controller failover and new topic creation.

ambroff replied:
Dang, you're right. That seems like a problem in this DynamicConfigManager thing. I just tested this and you're right, this code path is only exercised on first boot and when configuration is altered.

I think ideally the ConfigHandler would be invoked for topic creation as well.

ambroff replied:
I see. So DynamicConfigManager watches for entries in the zookeeper queue under `/config/changes/config_change_<sequence_number>`. That notification looks like this:

```
get /config/changes/config_change_0000000000
{"version":2,"entity_path":"topics/test-topic-2"}
```

So the simplest change would be to just have the controller add a notification to this queue for topic creation in addition to config alter.

I don't see a downside to doing that at the moment but I haven't tested it.

ambroff replied:
Oh this is weirder than I thought. That znode only appears to be written to by AdminZkClient :(. I wonder how it works with newer APIs

ambroff replied:
I pushed c4700b5 which I'm not totally sure is correct yet. It does seem to work properly when testing locally so far. It basically just performs the same notification to that changes znode when the broker creates a topic configuration znode.

gitlw replied:
My concern with doing this on the client side is that sometimes people use newer versions of open source clients without this path or obsolete client versions.
Thus I think it's safer to make this change on the server side, by having the controller to query the topic's configs during controller failover + when a new topic is created (besides using the dynamic config change mechanism).

ambroff replied:
Yeah :(. Initially I thought it would be OK to go that route since that code path is used by the handler of CreateTopicsRequest, but you're right we have many users at LinkedIn that don't use that yet.

I'll try the approach you suggested.

ambroff replied:
OK I pushed 071f7ed22 which I think handles this correctly.

ambroff committed:
Address some feedback from Lucas:

* Remove unused LiControlledShutdownSkipSafetyCheckResponse#parse
* Minor fixes
* Change a few places where the same container transformation happened
  for each iteration to just do it once.
* Fix confusing debug message.

ambroff committed:
Feedback from Lucas

Use mutable maps for shuttingDownBrokerIds and skipShutdownSafetyCheck
in ControllerContext

ambroff committed:
Feedback from Lucas

Remove unused LiControlledShutdownSkipSafetyCheckResponse#parse

ambroff committed:
Feedback from Lucas

Batch request in KakfaZkClient#getBrokerShutdownEntries

ambroff committed:
Feedback from Lucas

Batch broker shutdown znode removal.

ambroff committed:
Feedback from Lucas

Lucas pointed out that DynamicConfigManager will not pick up newly
created topics, only topic configuration changes.

I'm not sure if this is correct yet, but this patch just piggybacks on
the same notification mechanism to trigger DynamicConfigManager to
fetch topic configuration for newly created topics.

ambroff committed:
Fix test Lucas pointed out might be flaky.

ambroff committed:
Feedback from Lucas

Adding a comment explaining why the epoch of broker 2 is always 49 in
this test.

ambroff committed:
Revert "Feedback from Lucas"

This reverts commit c4700b57a232bcac69bc7ec465361ed12ca88da7.

ambroff committed:
Handle newly created topics

DynamicConfigManager will notify us for pre-existing topics on
startup, and for topic configuratin changes, but not newly created
topics. This covers that edge case.

ambroff committed:
Load all topic configs on controller failover

This picks up configuration for topics which were created between the
time the broker started up and controller failover happened.
